scripts/nn/layers/graph_conv.dml

```diff
@@ -0,0 +1,262 @@
+     *                This should be all 1s if there should be no edge weights.
+     * - W: linear weights, matrix of shape (f_in, f_out).
+     * - b: bias, matrix of shape (1, f_out).
+     * - add_self_loops: boolean that specifies whether self loops should be added.
+     *                   If TRUE new self loops will be added only for nodes that do
+     *                   not yet have a self loop. Added self loops will have weight 1.
+     *
+     * Outputs:
+     * - X_out: convolved and transformed node features, matrix of shape (n, f_out).
+     */
+    n = nrow(X)
+    m = nrow(edge_index)
+
+    # transform
+    X_hat = X %*% W
+
+    # count degrees for normalization
+    if (add_self_loops)
+    {
+        # add self loops and count degrees
+        D_in = matrix(1, n, 1)
+
+        loop_index = seq(1,n,1) %*% matrix(1, rows=1, cols=2)
+        loop_weight = matrix(1, rows=n, cols=1)
+
+        for (j in 1:m)
+        {
+            # count degree according to the weight for correct normalization
+            D_in[as.integer(as.scalar(edge_index[j, 2])), 1] += edge_weight[j, 1]
+
+            # for self loops
+            if (as.scalar(edge_index[j, 1]) == as.scalar(edge_index[j, 2]))
+            {
+                if (as.scalar(loop_weight[as.integer(as.scalar(edge_index[j, 1]))]) != 0.0)
+                {
+                    loop_weight[as.integer(as.scalar(edge_index[j, 1]))] = 0
+                    # remove 1 degree again that was added by initializing D_in with 1
+                    D_in[as.integer(as.scalar(edge_index[j, 2])), 1] += -1.0
+                }
+            }
+        }
+
+        edge_index = rbind(edge_index, loop_index)
+        edge_weight = rbind(edge_weight, loop_weight)
+    }
+    else
+    {
+        # count degrees
+        D_in = matrix(0, n, 1)
+        for (j in 1:m)
+        {
+            # count degree according to the weight for correct normalization
+            node_in = as.integer(as.scalar(edge_index[j, 2]))
+            D_in[node_in, 1] += edge_weight[j, 1]
+        }
+    }
+
+
+    # message passing/convolve with A_hat * X_hat (A_hat: normalized adjacency matrix weights)
+    X_out = matrix(0, rows=nrow(X_hat), cols=ncol(X_hat))
+    m = nrow(edge_index)
+    for (j in 1:m)
+    {
+        # doing: edge_weight[j] *= 1/sqrt(degree(node_in)*degree(node_out))
+        if (as.scalar(D_in[as.integer(as.scalar(edge_index[j, 2])), 1] *
+            D_in[as.integer(as.scalar(edge_index[j, 1])), 1]) != 0)
+        {
+            edge_weight[j, 1] = edge_weight[j, 1] * (1 / sqrt(D_in[as.integer(as.scalar(edge_index[j, 2])), 1]*
+                                                              D_in[as.integer(as.scalar(edge_index[j, 1])), 1]))
+        }
+        else
+        {
+            edge_weight[j, 1] = 0.0
+        }
+        X_out[as.integer(as.scalar(edge_index[j, 2]))] += as.scalar(edge_weight[j, 1]) * X_hat[as.integer(as.scalar(edge_index[j, 1]))]
```

Comment:
all these as.integer(as.scalar(...)) should not be needed.

Reply:
I think I got some issues when not using them. I will take a look at them.